update: add about defining Kotlin compiler execution strategy (#2844)
JetBrains/kotlin-web-site: docs/topics/gradle.md

  >
  {type="note"}
* If the `Xmx` is not specified, the Kotlin daemon will inherit it from the Gradle daemon.

## Defining Kotlin compiler execution strategy

> _Kotlin compiler execution strategy_ is a set of characteristics describing where Kotlin compiler is executed and some other nuances.

Reviewer: ```suggestion
_Kotlin compiler execution strategy_ defines a set of characteristics describing where Kotlin compiler is executed and some other nuances.
```

> _Kotlin compiler execution strategy_ defines a set of characteristics describing where Kotlin compiler is executed and some other nuances.

Author: Agree about vague.

There are three compiler execution strategies:

| Strategy       | Where Kotlin compiler is executed | Support for incremental compilation | Other characteristics                                                |
|----------------|-----------------------------------|-------------------------------------|----------------------------------------------------------------------|
| Daemon         | Inside its own daemon process     | Yes                                 | *Enabled by default*. Can be shared between different Gradle daemons |

> | In process     | Inside Gradle daemon process      | No                                  | May share heap with the Gradle daemon                                |

Reviewer: ```suggestion
| In process     | Inside the Gradle daemon process      | No                                  | May share heap with the Gradle daemon                                |
```

> | Out of process | In separate forked process        | No                                  | -                                                                    |

Reviewer: ```suggestion
| Out of process | In a separate forked process        | No                                  | -                                                                    |
```

Before Kotlin 1.6.20 to define a Kotlin compiler execution strategy, you used the system property `-Dkotlin.compiler.execution.strategy`.
This property may have been inconvenient in some cases. Kotlin 1.6.20 introduces the Gradle property with the same name `kotlin.compiler.execution.strategy`,
and the compile task property `compilerExecutionStrategy`.
The system property still works, but it will be removed in future releases. The priority of properties is the following:
* The task property `compilerExecutionStrategy` takes priority over the system and the Gradle properties `kotlin.compiler.execution.strategy`.
* The Gradle property takes priority over the system property.

The available values for `kotlin.compiler.execution.strategy` properties (both system and Gradle's) are:
1. `in-process`
2. `ouf-of-process`
3. `daemon` (default)

The available values for the `compilerExecutionStrategy` task property are:

> 1. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.IN_PROCESS`

Author: ```suggestion
1. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.DAEMON` (default)
```

2. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.OUT_OF_PROCESS`

> 3. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.DAEMON` (default)

Reviewer 2: Probably swap it with 1. position

Use the Gradle property `kotlin.compiler.execution.strategy` in `gradle.properties`:

```properties
kotlin.compiler.execution.strategy=out-of-process
```

Use the task property `compilerExecutionStrategy` in the `build.gradle.kts` buildscript:

```kotlin
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
import org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy

> // Some code

Reviewer 3: No "some code" please. Either check how we show Gradle scripts with imports in other examples on this page or consider using FQNs in the code.

  >
  {type="note"}
* If the `Xmx` is not specified, the Kotlin daemon will inherit it from the Gradle daemon.

## Defining Kotlin compiler execution strategy

> _Kotlin compiler execution strategy_ describes where Kotlin compiler is executed and if incremental compilation is supported in each case.

Reviewer 3: Maybe `defines`? Though the current wording seems OK too.

Author: Defines is better indeed, thank you.

> _Kotlin compiler execution strategy_ describes where Kotlin compiler is executed and if incremental compilation is supported in each case.

Author: ```suggestion
_Kotlin compiler execution strategy_ defines where Kotlin compiler is executed and if incremental compilation is supported in each case.
```

There are three compiler execution strategies:

| Strategy       | Where Kotlin compiler is executed | Support for incremental compilation | Other characteristics                                                  |
|----------------|-----------------------------------|-------------------------------------|------------------------------------------------------------------------|
| Daemon         | Inside its own daemon process     | Yes                                 | *The default strategy*. Can be shared between different Gradle daemons |
| In process     | Inside the Gradle daemon process  | No                                  | May share heap with the Gradle daemon                                  |

> | Out of process | In a separate forked process      | No                                  | -                                                                      |

Reviewer 3: Not sure if `forked` adds meaning here. Does it have any practical use?

Reviewer 3: For me, it only raised questions "Forked from what? What should I do with the fact that it was forked?"

Author: @ALikhachev could you help, please?

Reviewer: Yep, I guess we can remove `forked`. It's a technical detail
I'd suggest to rephrase it to be more clear
like
```suggestion
| Out of process | In a separate process for each call      | No                                  | -                                                                      |
```
So the new description emphasises that a new process will be created for each compilation task.

To define a Kotlin compiler execution strategy, you can use one of the following properties:
* The `kotlin.compiler.execution.strategy` Gradle property.
* The `compilerExecutionStrategy` compile task property.

> * The `-Dkotlin.compiler.execution.strategy` system property, which will be removed in future releases. 

Reviewer 3: Let's put the last thought in a separate note. Also, is it deprecated in 1.6.20? If yes, say this explicitly.

Author: >is it deprecated in 1.6.20?

I don't have such info, @ALikhachev could you help, please?

Reviewer: It's not deprecated yet, and we don't show any warnings on it's usages. We will add deprecation warning in 1.7 and then remove it in 1.8.

1. `daemon` (default)
2. `in-process`
3. `ouf-of-process`

Use the Gradle property `kotlin.compiler.execution.strategy` in `gradle.properties`:

```properties
kotlin.compiler.execution.strategy=out-of-process
```

The available values for the `compilerExecutionStrategy` task property are:
1. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.DAEMON` (default)
2. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.IN_PROCESS`
3. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.OUT_OF_PROCESS`

Use the task property `compilerExecutionStrategy` in your buildscripts:

<tabs group="build-script">
<tab title="Kotlin" group-key="kotlin">

```kotlin
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
import org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy

// ...

tasks.withType<KotlinCompile>().configureEach {
    compilerExecutionStrategy.set(KotlinCompilerExecutionStrategy.IN_PROCESS)
} 
```

</tab>
<tab title="Groovy" group-key="groovy">

```groovy
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
import org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy

// ...

> tasks.withType(KotlinCompile.class)

Author: ```suggestion
tasks.withType(KotlinCompile)
```
```suggestion
tasks.withType(KotlinCompile.class)
```
